cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ApplicationLifecycleService.java

```diff
@@ -345,6 +362,160 @@ public ApplicationWithPrograms updateApp(ApplicationId appId, AppRequest appRequ
+                                                             @Nullable String appVersion,
+                                                             @Nullable String currentConfigStr,
+                                                             ProgramTerminator programTerminator,
+                                                             ArtifactDetail artifactDetail,
+                                                             List<ApplicationConfigUpdateAction> updateActions,
+                                                             @Nullable KerberosPrincipalId ownerPrincipal,
+                                                             boolean updateSchedules) throws Exception {
+    ApplicationClass appClass = Iterables.getFirst(artifactDetail.getMeta().getClasses().getApps(), null);
+    if (appClass == null) {
+      throw new InvalidArtifactException(String.format("No application class found in artifact '%s' in namespace '%s'.",
+                                         artifactDetail.getDescriptor().getArtifactId(), namespaceId));
+    }
+    io.cdap.cdap.proto.id.ArtifactId artifactId =
+        Artifacts.toProtoArtifactId(namespaceId, artifactDetail.getDescriptor().getArtifactId());
+    EntityImpersonator classLoaderImpersonator = new EntityImpersonator(artifactId, this.impersonator);
+    ClassLoader artifactClassLoader =
+        artifactRepository.createArtifactClassLoader(artifactDetail.getDescriptor().getLocation(),
+                                                     classLoaderImpersonator);
+
+    String updatedAppConfig = "";
+    DefaultApplicationUpdateContext updateContext =
+        new DefaultApplicationUpdateContext(namespaceId, applicationId, artifactDetail.getDescriptor().getArtifactId(),
+                                            artifactRepository, currentConfigStr, updateActions);
+
+    // Run config update logic for the application to generate updated config.
+    try {
+      Object appMain = artifactClassLoader.loadClass(appClass.getClassName()).newInstance();
+      if (!(appMain instanceof Application)) {
+        throw new IllegalStateException(
+            String.format("Application main class is of invalid type: %s",
+                          appMain.getClass().getName()));
+      }
+      Application app = (Application) appMain;
+      Type configType = Artifacts.getConfigType(app.getClass());
+      try {
+        ApplicationUpdateResult<?> updateResult = app.updateConfig(updateContext);
+        updatedAppConfig = GSON.toJson(updateResult.getNewConfig(), configType);
+      } catch (UnsupportedOperationException ex) {
+        String errorMessage = String.format("application of type %s does not support update.",
+                                            appMain.getClass().getName());
+        LOG.error("Application update failed due to " + errorMessage);
+        throw ex;
+      }
+    } catch (Exception ex) {
```

Comment:
what are the different exception that can be thrown here? Not all exceptions have a cause